### PR TITLE
[SYCL][libdevice] Split MSVC specific math functions from cmath devicelib.

### PR DESCRIPTION
**Pull request by jinge90**

### CODE REVIEW

**Comment by jinge90** [comment]
In cmath devicelib, we disabled ldexpf, hypotf, frexpf for Windows, the reason was we included "math.h" in the source and MSVC "math.h" has inlined those 3 functions in header file which would block us to implement those 3 functions, otherwise "multiple definition" error would happen in building cmath devicelib.
The reason why we had to include MSVC "math.h" was we implemented some MSVC internal math functions in cmath devicelib and those internal math functions depended on some constant which are defined in MSVC math header.
Now, some users reported that they encountered "undefined symbol" error for "hypotf" on SYCL GPU device when using 3rd-party math headers instead of MSVC math headers. In order to support such use case, we need to provide ldexpf, hypotf, frexpf in SYCL device library for Windows too.
And in order to resolve the multiple definition error introduced by inline for hypotf, ldexpf, frexpf in MSVC math header, we have to:
1. split all MSVC internal function into a new source file msvc-math.cpp and we will include "math.h" in this source.
2. enable hypotf, ldexpf, frexpf in cmath_wrapper.cpp since there is no "include <math.h>" in the source any more, building it using clang will succeed.
3. we will build msvc-math.cpp to get a new .obj file libsycl-msvc-math.obj and this devicelib will be added to device link list by default and controlled by fsycl-device-lib=libm-fp32 too.


Thanks very much.

**Comment by jinge90** [comment]
Hi, @mdtoguchi 
Could you help review the driver and driver lit test for this PR?

Thanks very much.

File: clang/lib/Driver/ToolChains/SYCL.cpp [hunk 1]
```diff
@@ -130,21 +130,16 @@ void SYCL::constructLLVMForeachCommand(Compilation &C, const JobAction &JA,
...
-    "fallback-cmath",
-    "fallback-cmath-fp64",
-    "fallback-complex",
-    "fallback-complex-fp64"};
+
+static llvm::SmallVector<StringRef, 16> SYCLDeviceLibList {
+  "crt", "cmath", "cmath-fp64", "complex", "complex-fp64",
+#if defined(_WIN32)
+      "msvc-math",
+#endif
```
Comment (mdtoguchi) [inline clang/lib/Driver/ToolChains/SYCL.cpp:138]: Indentation is off, I would assume that clang-format forced this behavior?

Comment (jinge90) [reply clang/lib/Driver/ToolChains/SYCL.cpp:138]: Yes, I ran clang-format before submitting the patch.

File: libdevice/cmath_wrapper.cpp [hunk 2]
```diff
@@ -36,20 +36,12 @@ float logf(float x) { return __devicelib_logf(x); }
...
 DEVICE_EXTERN_C_INLINE
 float expf(float x) { return __devicelib_expf(x); }
-// On Windows, the math.h includes wrapper definition for functions:
-// frexpf, ldexpf, hypotf, so we can't provide these 3 functions in
-// device libraries which may lead to multiple definition error.
-// The "hypotf" on Windows will call an internal function "_hypotf"
-// and frexpf, ldexpf will call corresponding double version:
-// frexp and ldexp. frexpf and ldexpf can only be used on platforms
-// with fp64 support currently.
-#ifndef _WIN32
```
Comment (steffenlarsen) [inline libdevice/cmath_wrapper.cpp:46]: Why is this guard no longer needed?

Comment (steffenlarsen) [reply libdevice/cmath_wrapper.cpp:46]: Sorry, this seems to be described in a previous comment.

**Review by steffenlarsen — APPROVED** [review]
LGTM!

**Review by pvchupin — APPROVED** [review]

**Pull request merged**